src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -634,6 +654,9 @@ public boolean hasntStartedYet() {
 
     @RequirePOST
     public synchronized HttpResponse doStop() {
+        if (hasPermission(Item.CANCEL)) {
+            stopCalled = true;
```

Comment:
Possibly better to move this [here](https://github.com/jenkinsci/workflow-job-plugin/blob/ba83d8f854b6befcb64c1b824ecdef207658b369/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java#L238).